test/extensions/filters/network/ssh/service_userauth_test.cc

```diff
@@ -0,0 +1,746 @@
+  auto r = service_->handleMessage(std::move(msg));
+  ASSERT_EQ(r.code(), absl::StatusCode::kInternal);
+  ASSERT_EQ(r.message(), "invalid target");
+}
+
+TEST_F(DownstreamUserAuthServiceTest, HandleMessageServerDeny) {
+  auto server_msg = std::make_unique<pomerium::extensions::ssh::ServerMessage>();
+  auto deny = server_msg->mutable_auth_response()->mutable_deny();
+  deny->set_partial(true);
+  deny->add_methods("publickey");
+  deny->add_methods("keyboard-interactive");
+
+  wire::Message ssh_msg;
+  EXPECT_CALL(*transport_, sendMessageToConnection(_))
+    .WillOnce(DoAll(SaveArg<0>(&ssh_msg),
+                Return(absl::UnknownError("sentinel"))));
+
+  auto r = service_->handleMessage(std::move(server_msg));
+  ASSERT_EQ(r, absl::UnknownError("sentinel"));
+
+  ASSERT_EQ(wire::SshMessageType::UserAuthFailure, ssh_msg.msg_type());
+  auto failure_msg = ssh_msg.message.get<wire::UserAuthFailureMsg>();
+  ASSERT_TRUE(failure_msg.partial);
+  ASSERT_EQ((string_list{"publickey", "keyboard-interactive"}), failure_msg.methods);
+}
+
+TEST_F(DownstreamUserAuthServiceTest, HandleMessageServerDenyNoMethods) {
+  auto msg = std::make_unique<pomerium::extensions::ssh::ServerMessage>();
+  msg->mutable_auth_response()->mutable_deny();
+  auto r = service_->handleMessage(std::move(msg));
+  ASSERT_EQ(r.code(), absl::StatusCode::kPermissionDenied);
+  ASSERT_EQ(r.message(), "");
+}
+
+TEST_F(DownstreamUserAuthServiceTest, HandleMessageServerInfoRequest) {
+  auto server_msg = std::make_unique<pomerium::extensions::ssh::ServerMessage>();
+  auto req = server_msg->mutable_auth_response()->mutable_info_request();
+  req->set_method("keyboard-interactive");
+  pomerium::extensions::ssh::KeyboardInteractiveInfoPrompts prompts{};
+  prompts.set_name("prompts-name");
+  prompts.set_instruction("prompts-instruction");
+  auto prompt1 = prompts.add_prompts();
+  prompt1->set_echo(true);
+  prompt1->set_prompt("username");
+  auto prompt2 = prompts.add_prompts();
+  prompt2->set_prompt("password");
+  req->mutable_request()->PackFrom(prompts);
+
+  wire::Message ssh_msg;
+  EXPECT_CALL(*transport_, sendMessageToConnection(_))
+    .WillOnce(DoAll(SaveArg<0>(&ssh_msg),
+                Return(absl::UnknownError("sentinel"))));
+
+  auto r = service_->handleMessage(std::move(server_msg));
+  ASSERT_EQ(r, absl::UnknownError("sentinel"));
+
+  ASSERT_EQ(wire::SshMessageType::UserAuthInfoRequest, ssh_msg.msg_type());
+  auto request_msg =
+    ssh_msg.message.get<wire::detail::overload_set_for_t<wire::UserAuthInfoRequestMsg>>()
+      .resolve<wire::UserAuthInfoRequestMsg>()->get();
+  ASSERT_EQ("prompts-name", request_msg.name);
+  ASSERT_EQ("prompts-instruction", request_msg.instruction);
+  ASSERT_EQ(2, request_msg.prompts->size());
+  ASSERT_TRUE(request_msg.prompts[0].echo);
+  ASSERT_EQ("username", request_msg.prompts[0].prompt);
+  ASSERT_FALSE(request_msg.prompts[1].echo);
+  ASSERT_EQ("password", request_msg.prompts[1].prompt);
```

Comment:
This can be simplified:
```cc
  auto matcher = MSG(wire::UserAuthInfoRequestMsg,
                     AllOf(FIELD_EQ(name, "prompts-name"),
                           FIELD_EQ(instruction, "prompts-instruction"),
                           FIELD_EQ(prompts,
                                    testing::ElementsAre(
                                      wire::UserAuthInfoPrompt{.prompt = "username"s, .echo = true},
                                      wire::UserAuthInfoPrompt{.prompt = "password"s, .echo = false}))));
  EXPECT_CALL(*transport_, sendMessageToConnection(matcher))
    .WillOnce(Return(absl::UnknownError("sentinel")));

  auto r = service_->handleMessage(std::move(server_msg));
  ASSERT_EQ(r, absl::UnknownError("sentinel"));
}
```